crates/core/src/db/commit_log.rs

```diff
@@ -25,27 +25,236 @@ use std::io;
+    /// and can thus safely be written to via the resulting [`CommitLogMut`].
+    pub fn replay<F>(&self, mut f: F) -> Result<CommitLogMut, DBError>
+    where
+        // TODO(kim): `&dyn ObjectDB` should suffice
+        F: FnMut(Commit, Arc<Mutex<Box<dyn ObjectDB + Send>>>) -> Result<(), DBError>,
+    {
+        let unwritten_commit = {
+            let mut mlog = self.mlog.lock().unwrap();
+            let total_segments = mlog.total_segments();
+            let segments = mlog.segments();
+            let mut iter = Replay {
+                tx_offset: 0,
+                last_commit_offset: None,
+                last_hash: None,
+
+                segments,
+                segment_offset: 0,
+                current_segment: None,
+            };
+
+            for commit in &mut iter {
+                match commit {
+                    Ok(commit) => f(commit, self.odb.clone())?,
+                    Err(ReplayError::Other { source }) => return Err(source.into()),
+
+                    // We expect that partial writes can occur at the end of a
+                    // segment. Trimming the log is, however, only safe if we're
+                    // at the end of the _log_.
+                    Err(ReplayError::OutOfOrder {
+                        segment_offset,
+                        last_commit_offset,
+                        decoded_commit_offset,
+                        expected,
+                    }) if segment_offset < total_segments - 1 => {
+                        log::warn!("Out-of-order commit {}, expected {}", decoded_commit_offset, expected);
+                        return Err(LogReplayError::TrailingSegments {
+                            segment_offset,
+                            total_segments,
+                            commit_offset: last_commit_offset,
+                            source: io::Error::new(io::ErrorKind::Other, "Out-of-order commit"),
+                        }
+                        .into());
+                    }
+                    Err(ReplayError::CorruptedData {
+                        segment_offset,
+                        last_commit_offset: commit_offset,
+                        source,
+                    }) if segment_offset < total_segments - 1 => {
+                        log::warn!("Corrupt commit after offset {}", commit_offset);
+                        return Err(LogReplayError::TrailingSegments {
+                            segment_offset,
+                            total_segments,
+                            commit_offset,
+                            source,
+                        }
+                        .into());
+                    }
+
+                    // We are near the end of the log, so trim it to the known-
+                    // good prefix.
+                    Err(
+                        ReplayError::OutOfOrder { last_commit_offset, .. }
+                        | ReplayError::CorruptedData { last_commit_offset, .. },
```

Comment:
We can't know if it is the last commit exactly -- `MessageLog::open_segment_max_offset` (meaning: total number of entries) is calculated without decoding the payload or checking integrity, it only means that the number of bytes is aligned.

Since commits are variable-length, we could only apply a heuristic such as "not more than ~50KiB remaining".

A viable approach for a more robust solution might be to calculate a CRC32 for every record, and allow to skip over out-of-order commits _iff_ the next commit has the right sequence number (although I don't know how this could arise).